### PR TITLE
First run at a native html backend for asciimath.rb.

### PR DESCRIPTION
**Pull request by Tobski**
I've been using asciimath as part of asciidoctor, and the large technical doc we're generating is resulting in enormous slowdown when using JS to process mathml output, so we've resorted to not using math notation for several things. We'd rather avoid the cost of the JS and associated paraphenalia, which is why I've looked into doing this.

It seems MathML has been abandoned as a web standard, as a couple of browsers have cited insecurity. However there's a bunch of layouts in CSS3 which have made doing a direct-to-html parser quite reasonable, which is what I've tried to do here.

There's still a few bits TODO, but it's mostly workable in its current form IMO, so thought I'd raise a PR to see if you had any feedback, and to see if you'd consider adopting this sort of thing into your asciimath gem, or if you'd rather it stay as a forked project?

### CODE REVIEW

**Comment by pepijnve** [comment]
@Tobski thanks for the contribution. I'll have a more in depth look at the code later today. At first glance it looks good. I more than happy to integrate this into the project; no need to maintain a fork (unless you would prefer that yourself of course).

This project is/was actually intended to become part of the asciidoctor organisation, but that stalled for no particular reason whatsoever. I'll pick that up with @mojavelinux.

The logical next step for this PR will be to integrate it into the asciidoctor HTML5 backend. That's part of the asciidoctor project so we'll need a PR there once a new release of the asciimath gem has been built.

**Comment by TobiasHector** [comment]
@pepijnve Awesome! More than happy to have this thing integrated directly into the core project, let me know how you want to proceed once you've done a deeper review 😄

**Comment by Tobski** [comment]
@pepijnve Awesome! More than happy to have this thing integrated directly into the core project, let me know how you want to proceed once you've done a deeper review 😄

File: lib/asciimath/html.rb [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                subsup(expression[:s1],expression[:s2],expression[:s3])
+              elsif operator == :underover
+                # TODO: Handle over/under braces in some way? SVG maybe?
+                underover(expression[:s1],expression[:s2],expression[:s3])
+              end
+            when :matrix
+              self.row do
+                # TODO: Figure out how big the brace should be and insert something to that effect...
+                self.brace(expression[:lparen]) if expression[:lparen]
+                matrixWidth  = "grid-template-columns:repeat(" + expression[:rows][0].length.to_s + ",1fr);"
```
Comment (pepijnve) [inline lib/asciimath/html.rb]: use snake case for variable names

File: lib/asciimath/html.rb [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        end
+      end
+    end
+
+    private
+
+    def append(expression, opts = {})
+      case expression
+        when Array
+          self.row do
```
Comment (pepijnve) [inline lib/asciimath/html.rb]: I would remove the `self.`s throughout the code to be consistent with the rest.

**Comment by Tobski** [comment]
I've fixed the cosmetic bits, I'll add some unit tests and get back to you.

**Comment by pepijnve** [comment]
Thanks. The unit tests I currently have in place are simple black box tests that verify expected output for a set of inputs. Something like that is good enough for the HTML code as well. It's just so there's at least something in place that automatically runs the code on travis to avoid regressions or unintended changes to the output.

**Comment by Tobski** [comment]
Hey, sorry I went silent there for a few days - have had a hectic weekend!

So I thought to myself "well I'll just use the unit tests you already have rather than formalise the ones I do have", and of course, it turned up an issue.

So I've half-punted on braces, by currently "just" putting them at a set size, I haven't yet figured out a way to automatically size them in the html, though I have some ideas. For now though, making them always double height makes the thing work for now - just not ideally.

However it turns out I have the same issue for the square root function, which I hadn't spotted in the code previously, so hadn't thought to handle it. Unfortunately since that's not "just" a simple character add, I am unsure what to do with it.

The same ideas that might work for braces *should* work for square root too - but it'll take time to experiment and implement those.

I'm happy to either go off and fix that before merging, or I can just get the rest of the unit tests together (and hope there's nothing else glaringly wrong :D) and we can call it "experimental" until that gets implemented properly?

What do you think?

**Comment by pepijnve** [comment]
> we can call it "experimental" until that gets implemented properly

That works for me. It doesn't have to be 100% perfect to already be useful for people.

**Comment by Tobski** [comment]
Ok I went through all your test cases and got what I could rendering, fixed a few things along the way, and added some form of documentation to the README (including known issues).

Seem ok to you?

**Comment by Tobski** [comment]
Oh hm. Looks like the "unsupported" tests are causing build failures, I'll skip said tests for now.

**Comment by Tobski** [comment]
I added some fixes for a few other "simple" things I could do, but I'm going to leave it where it is for now barring public outcry :)

**Comment by pepijnve** [comment]
I'll try to have a look asap and get this merged. I can then make a new release of the gem and see if we can get this hooked into asciidoctor as an optional math processing option.

**Comment by TobiasHector** [comment]
Hey @pepijnve did you have any chance to look at this at all yet? Just remembered I still had this outstanding so thought I'd come check up on it :)

There's no urgency on my part, would be nice to have it in place early next year though so I can work on the next steps I need after this is done.

**Comment by pepijnve** [comment]
Not yet I'm afraid. There's nothing blocking this except for availability of spare time to spend on this. Sorry for the delay.

**Comment by TobiasHector** [comment]
No problem, if I end up getting blocked on this I can always just incorporate the changes for our project locally - just thought I'd ping on it whilst I remembered!

**Comment by Tobski** [comment]
@pepijnve I just remembered this was a thing - at some point I'd like to revive this and get back to work on it as I believe it will be quite useful. Do you think you'll have any time to look at it in the near future? If not that's fine, just trying to set expectations!

**Comment by Tobski** [comment]
Or maybe I should ping @mojavelinux at this point since ownership has been transferred... 🤔 ?

**Comment by pepijnve** [comment]
I forgot about this one myself as well. There's some other projects taking up my spare open source dev time at the moment. Not sure when I'll get round to this.

**Comment by Tobski** [comment]
Shoot, that's annoying. I'm not on any deadline at the moment though, so I'll just have to figure out what I want to do in the meantime.

**Comment by pepijnve** [comment]
I had a quick look at the changes again. Looks good to me, so I'm going to go ahead and merge this.
Still won't be automatically integrated into the asciidoctor HTML5 backend though.

**Comment by Tobski** [comment]
Oh amazing, thanks 😄 - yea I'll look into the asciidoctor backend when I get a spare minute, just having this integrated is a great step.

Is there any likelihood of a new release of the gem any time soon that'll capture this change? I have no idea how much work that involves so no idea if that's a big ask or not 🤔

**Comment by pepijnve** [comment]
Not much work at all. Give me a minute 😄

**Comment by pepijnve** [comment]
1.0.5 is available on rubygems.org now

**Comment by Tobski** [comment]
❤️

**Pull request merged**